### ISSUE TITLE
Build Failed; target 'clean' does not exist under \FDA-HPHI-UserRegWS-master\build.xml 

### ISSUE THREAD

**Issue by zardarkhan** [PopMedNet-Team/Mobile-Application#7]
I'm following steps from FDA_WCP_UR_AppSetup.

I've installed all the prerequisites, setup the labkey development server. I'm stuck on _Step 5.2; User Registration Web Services_.

When I run 'ant clean' in the 'C:\dev\labkey\trunk\server\customModules\FDA-HPHI-UserRegWS-master\' folder - I get the following error:
```
BUILD FAILED
Target "clean" does not exist in the project "Build FdahpUserRegWS".

Total time: 0 seconds
```
Any ideas what might cause this error? I've noticed some sections for the build.xml file have been commented off - I've tried uncommenting them but no luck. Here's the build.xml itself:

```xml
<?xml version="1.0" encoding="ISO-8859-1"?>
<!--
<project name="Build FdahpUserRegWS" default="build" basedir="../../">
-->
<project name="Build FdahpUserRegWS" default="build" basedir=".">
    <dirname property="moduleDir" file="${ant.file}"/>

   <!-- <path id="classpath">
        <pathelement location="lib/*.jar"/>
        &lt;!&ndash; <pathelement location="lib/jython.jar"/> &ndash;&gt;
        &lt;!&ndash; <pathelement location="lib/javax.servlet.jar"/> &ndash;&gt;
    </path>-->
    <!-- TODO: Need to set some properties for this to work standalone -->
    <target name="init">
    </target>

    <target name="build" depends="init, build_module_api, build_module" />

    <!-- Called by the main build.xml with inheritall = true; other callers (e.g., top-level build target)
         must define all the properties used below. -->
    <target name="build_module_api">
        <ant antfile="${basedir}/build.xml"
             target="sub_build_module_api"
             inheritall="true">
        </ant>
    </target>

    <!-- Called by the main build.xml with inheritall = true; other callers (e.g., top-level build target)
         must define all the properties used below. -->
    <target name="build_module" description="Builds FdahpUserRegWS module and deploys into LabKey installation.">
        <ant antfile="${basedir}/build.xml"
             target="sub_build_module"
             inheritall="true">
        </ant>

      <!--  <classpath>
            <path refid="component.class.path"/>
            <pathelement path="${build.modules.dir}/flow/explodedModule/lib/schemas.jar"/>
        </classpath>-->
        <!-- Compile the FdahpUserRegWS library JARs -->
        <!--<javac srcdir="${modules.dir}/fdahpUserRegWS/src" destdir="${build.dir}/fdahpUserRegWS/classes">
            <classpath>
                <pathelement path="${build.dir}/fdahpUserRegWS/lib/*.jar"/>
            </classpath>
        </javac>-->
    </target>
    <!--
    <target name="compile">
        <mkdir dir="./build/classes"/>
        <javac srcdir="./src" destdir="./build/classes"/>
    </target>
    -->
</project>
```

**Comment by s37-robert** [comment]
I seem to be in the same boat as you are @zardarkhan . The most recent version of LabKey is built using Gradle/IntelliJ instead of Ant. Try adding the following line in settings.gradle and rebuilding the app by running `gradlew deployAll` from {LABKEY_HOME}:

```
include ":server:customModules:fdahpUserRegWS"
```

* Note that you may want to copy `FDA-HPHI-StudyDesignerWS-master` to `fdahpUserRegWS`. Not sure if the system can handle punctuation in the module name.

**Comment by clemens** [comment]
@s37-robert May I ask: How did you get it running? I seem to have multiple issues:

- Encoding is set to ISO-8859-1 (but not everywhere, apparently) and gradlew refuses to compile. (I'm running macOS which defaults to UTF-8.)
- I get multiple`package javax.persistence does not exist warnings.

I'd be grateful for any pointers.

**Comment by zardarkhan** [comment]
> I seem to be in the same boat as you are @zardarkhan . The most recent version of LabKey is built using Gradle/IntelliJ instead of Ant. Try adding the following line in settings.gradle and rebuilding the app by running `gradlew deployAll` from {LABKEY_HOME}:
> 
> ```
> include ":server:customModules:fdahpUserRegWS"
> ```
> * Note that you may want to copy `FDA-HPHI-StudyDesignerWS-master` to `fdahpUserRegWS`. Not sure if the system can handle punctuation in the module name.

@s37-robert Thank you! I was able to update the settings.gradle and rebuild the app successfully. I couldn't get 'gradlew deployAll' to work but 'gradlew deployApp' ran smoothly. 

However, when I try to navigate to fdahpUserRegWS/ping.api over localhost - with LabKey Dev running in IntelliJ - I get the following message: "No LabKey Server module registered to handle request for controller: fdahpuserregws". Any ideas? 

I will investigate on my end and update with more information if successful.

**Comment by s37-robert** [comment]
Hi @zardarkhan, seems like you got stuck at the same place I did. The updated script built LabKey successfully and I was able to set up the server but the fdahpUserRegWS module was not installed. Digging a bit deeper, it turns out that module failed to build with errors in the following files:

In FdahpUserRegWSManager.java: There are numerous calls to TableInfo.setAuditBehavior(). This method was removed from the LabKey API in the latest version. I couldn't do anything other than comment out those lines.

In FdahpUserRegUtil.java: There is a dependency on com.notnoop.apns which cannot be resolved. For some reason, it's not resolving to the JAR file at lib/apns-1.0.0.Beta6.jar. The build.gradle file in customModules says that modules are built using the org.labkey.module plugin but I can't find the source to confirm whether it still looks for dependencies in the /lib folder.

At this point, I'm thinking that either the codebase needs to be updated for compatibility with the latest version of LabKey or the LabKey dependency needs to be pinned to the previous version. Building using the previous version is not without its drawbacks either since the public enlistment is missing the `registration` distribution and the `compliance` module.

Will update if I have more luck building this.

**Comment by s37-robert** [comment]
Update: I got the ping.api endpoint working using this line in settings.gradle instead of the one mentioned above:

```
BuildUtils.includeModules(this.settings, [":server:customModules:fdahpUserRegWS"])
```

**Comment by formanojhr** [comment]
@s37-robert  Thanks for the update. 

1. Did you run gradlew deployApp or gradlew deployAll. deployAll did not seem to run for me. It seems like it is not defined in the build file.  
2. Also can you please confirm your directory structure? I have the following :../ LabKey/trunk within which I have copied ../server/customModules/FDA-HPHI-UserRegWS-master and within it ../server/customModules/FDA-HPHI-UserRegWS-master/FDA-HPHI-StudyDesignerWS-master. Is that accurate based on your setup? 
3. And where do I confirm if the custom modules got picked up while deploying ? Is there a log file somewhere for gradle deploy? Thanks! 
4. I am also running the labKey server dev start in intellij. Is that the server that host the module? Or Deploy all also brings up another server ?

**Comment by zardarkhan** [comment]
> Update: I got the ping.api endpoint working using this line in settings.gradle instead of the one mentioned above:
> 
> ```
> BuildUtils.includeModules(this.settings, [":server:customModules:fdahpUserRegWS"])
> ```

Hey @s37-robert - thanks for all your help! 
I've tried updating settings.gradle with the `BuildUtils...` line followed by cleaning and rebuilding the project but I'm getting the same _No LabKey Server module registered to handle request for controller: fdahpuserregws_. I will dig around to see what may be causing this - will keep you posted.

Update: I saved the `gradlew deployApp` log to a file and noticed that `:server:customModule:` is skipped - see the txt file I've attached below:

[gradleLog.txt](https://github.com/PopMedNet-Team/FDA-My-Studies-Mobile-Application-System/files/2733925/gradleLog.txt)

**Comment by zardarkhan** [comment]
> @s37-robert Thanks for the update.
> 
> 1. Did you run gradlew deployApp or gradlew deployAll. deployAll did not seem to run for me. It seems like it is not defined in the build file.
> 2. Also can you please confirm your directory structure? I have the following :../ LabKey/trunk within which I have copied ../server/customModules/FDA-HPHI-UserRegWS-master and within it ../server/customModules/FDA-HPHI-UserRegWS-master/FDA-HPHI-StudyDesignerWS-master. Is that accurate based on your setup?
> 3. And where do I confirm if the custom modules got picked up while deploying ? Is there a log file somewhere for gradle deploy? Thanks!
> 4. I am also running the labKey server dev start in intellij. Is that the server that host the module? Or Deploy all also brings up another server ?

Hey @formanojhr, I can take a stab at a few of your questions:
1) `gradlew deployApp` will run for you. You can see gradlew tasks associated with this build by running `gradlew tasks`.
2) As I understand it - that is correct and this is exactly how I have my directory setup.
3) From what I've gathered, you can run `gradlew deployApp > gradleLog.txt 2>&1` which will output a txt file labelled _gradleLog.txt_ into your _server/trunk_ directory.
See this Stack Overflow post for more info: [here](https://stackoverflow.com/questions/35558637/where-are-gradle-logs)
4) I don't quite understand the last question, can you rephrase it?

**Comment by formanojhr** [comment]
HI @zardarkhan  Thanks to you as well. Thanks for all the answers.
So, when I enabled the build for the custom module and then ran ./gradlew clean build I get the following error:  Did you disable the ANT build.xml or something  ? Thanks! 

```
> >  Task :server:customModules:fdahpUserRegWS:ant_build_module_api FAILED
> 
> FAILURE: Build failed with an exception.
> 
> * What went wrong:
> Execution failed for task ':server:customModules:fdahpUserRegWS:ant_build_module_api'.
> > The following error occurred while executing this line:
>   java.io.FileNotFoundException: /Users/mramakrishnan/Documents/workspaces/IdeaProjects/LabKey/trunk/server/build.xml (No such file or directory)
 **Update on above:**
DO NOT run above command ; It kicks off ANT build which is not needed.  Look at this comprehensive list of gradle commands mapping to ant: 
https://www.labkey.org/Documentation/wiki-page.view?name=gradlebuild
```
As documented above, gradlew deployApp is the equivalent of ant build command. 

On questions 4. I do not understand where from the deployed custom modules gets picked up and hosted. I am following the LabKey dev setup as here:  https://www.labkey.org/Documentation/wiki-page.view?name=devMachine. And, running a localhost server from that start config on intellij. What I do not understand is if deployApp deploys and gets picked up from the LabKey server which runs on intellij or deployApp starts another Tomcat server by itself. From what I understand from Labkey doc, it just gets the artifact ready for deployment as per this doc: https://www.labkey.org/Documentation/wiki-page.view?name=buildLabKey#directories. Are you also running a dev setup from intellij for Labkey?

**Comment by formanojhr** [comment]
@zardarkhan  I am in the same stage as you ; Getting  the labkey module not deployed error:
No LabKey Server module registered to handle request for controller: fdahpuserregws
http://localhost:8090/labkey/fdahpuserregws/ping.api

@s37-robert  or @zardarkhan  Did you add a gradle build file to the fdahpuserregws project since when I run deployApp the project compile breaks with errors like this:
/dto/QuestionResponseSubTypeDto.java:28: error: package javax.persistence does not exist
import javax.persistence.Entity;
To me it seems like dependencies are not picked up for java compile. 
Love to get your input. we really need to update documentation on this one for other trying to get this running. Cheers!

**Comment by formanojhr** [comment]
@clemens  @zardarkhan  - I got the ping.api by building the project with the following command by adding a build.gradle file under the fdahpUserRegWS project directory root with the following content : 

import org.labkey.gradle.util.BuildUtils

apply plugin: 'java'
apply plugin: 'org.labkey.module'

repositories {
    flatDir {
        dirs 'lib'
    }
    mavenCentral()
}

sourceSets {
    engine {
        java {
            srcDirs = ['./enginesrc']
        }
    }
}

dependencies {
    compile name: 'apns-1.0.0.Beta6'
    compile group: 'org.labkey', name: 'labkey-client-api', version: '15.3.0'
    compile name: 'labkey-client-api-17.3'
    BuildUtils.addLabKeyDependency(project: project, config: "compile", depProjectPath: ":server:api")
    BuildUtils.addLabKeyDependency(project: project, config: "compile", depProjectPath: ":server:internal")
    engineCompile "org.labkey:labkey-client-api:${project.version}"
}

2. I also had to fix some code in FdahpUserRegWSManager.java to avoid the labkey API build errors:
TableInfo table = FdahpUserRegWSSchema.getInstance().<SomeMethod> to
SchemaTableInfo table = (SchemaTableInfo)FdahpUserRegWSSchema.getInstance().<SomeMethod>


3. Then run the command to build command for the module. 

./gradlew :server:customModules:fdahpUserRegWS:deployModule.  
Remember to add the following line to settings.gradle to include the custom module we added;
BuildUtils.includeModules(this.settings, [":server:customModules:fdahpUserRegWS"])

Another issue I ran into is LabKey server run all scripts under the resources/schema/<yourdb> directory when it finds the new war build. Since I had already run those scripts on postgres ( which I chose) there were lots 'Already Exists" errors. They are meant to be rerun. So, I moved the schema directory to avoid them from being run on the already existing db. 

Hope this helps.

**Comment by zardarkhan** [comment]
@formanojhr That's fantastic! Glad to hear you have it running now.

Unfortunately, I haven't had chance to work on deploying LabKey since my last post. Thank you for the pointers; I will attempt now and keep posted of my progress.

**Comment by zardarkhan** [comment]
@formanojhr 

First I followed the steps you mentioned above and I got the following error:

```
> Could not resolve all task dependencies for configuration ':server:customModules:FDA-HPHI-UserRegWS:compileClasspath'.
   > Could not find :labkey-client-api-17.3:.
     Searched in the following locations:
       - file:/C:/Program Files/Apache Software Foundation/Tomcat 8.5/lib/labkey-client-api-17.3.jar
       - file:/C:/Program Files/Apache Software Foundation/Tomcat 8.5/bin/labkey-client-api-17.3.jar
       - file:/C:/dev/labkey/trunk/server/customModules/FDA-HPHI-UserRegWS/lib/labkey-client-api-17.3.jar
     Required by:
         project :server:customModules:FDA-HPHI-UserRegWS
````
I noticed that there wasn't a 'labkey-client-api-17.3' wasn't located in the lib folder. I downloaded the .jar and reran. Getting the following error:

```
> Task :server:customModules:FDA-HPHI-UserRegWS:compileJava FAILED
C:\dev\labkey\trunk\server\customModules\FDA-HPHI-UserRegWS\src\com\hphc\mystudies\FdahpUserRegWSManager.java:111: error: cannot find symbol
            SchemaTableInfo table = (SchemaTableInfo)FdahpUserRegWsSchema.getInstance().getParticipantDetails();
            ^
  symbol:   class SchemaTableInfo
  location: class FdahpUserRegWSManager
```

I will keep everyone posted on this

**Comment by seanthingee** [comment]
@zardarkhan Hello, I'm also trying to set this up and having the same issues. I wanted to mention that I added: `import org.labkey.api.data.SchemaTableInfo;` to resolve the `cannot find symbol` and now I receive:
```
> Task :server:customModules:fdahpUserRegWS:compileJava FAILED
/home/centos/labkey/trunk/server/customModules/fdahpUserRegWS/src/com/hphc/mystudies/FdahpUserRegWSManager.java:110: error: incompatible types: FdahpUserRegWSSchema cannot be converted to SchemaTableInfo
                SchemaTableInfo table = (SchemaTableInfo)FdahpUserRegWSSchema.getInstance();
                                                                                         ^
```
Not much better.

Have you been able to make any progress since you last posted?

**Comment by zardarkhan** [comment]
> @zardarkhan Hello, I'm also trying to set this up and having the same issues. I wanted to mention that I added: `import org.labkey.api.data.SchemaTableInfo;` to resolve the `cannot find symbol` and now I receive:
> 
> ```
> > Task :server:customModules:fdahpUserRegWS:compileJava FAILED
> /home/centos/labkey/trunk/server/customModules/fdahpUserRegWS/src/com/hphc/mystudies/FdahpUserRegWSManager.java:110: error: incompatible types: FdahpUserRegWSSchema cannot be converted to SchemaTableInfo
>                 SchemaTableInfo table = (SchemaTableInfo)FdahpUserRegWSSchema.getInstance();
>                                                                                          ^
> ```
> Not much better.
> 
> Have you been able to make any progress since you last posted?

Sorry about the super-late reply. 

Unfortunately I won't be of much help. I have not made any progress since my last post and have put aside any efforts to get the system up and running. I hope you've had better luck. Try to document any success you've had because there's definitely more people running into the same issues as us.

Cheers.

**Comment by JonatanJJ** [comment]
> @formanojhr
> 
> First I followed the steps you mentioned above and I got the following error:
> 
> ```
> > Could not resolve all task dependencies for configuration ':server:customModules:FDA-HPHI-UserRegWS:compileClasspath'.
>    > Could not find :labkey-client-api-17.3:.
>      Searched in the following locations:
>        - file:/C:/Program Files/Apache Software Foundation/Tomcat 8.5/lib/labkey-client-api-17.3.jar
>        - file:/C:/Program Files/Apache Software Foundation/Tomcat 8.5/bin/labkey-client-api-17.3.jar
>        - file:/C:/dev/labkey/trunk/server/customModules/FDA-HPHI-UserRegWS/lib/labkey-client-api-17.3.jar
>      Required by:
>          project :server:customModules:FDA-HPHI-UserRegWS
> ```
> I noticed that there wasn't a 'labkey-client-api-17.3' wasn't located in the lib folder. I downloaded the .jar and reran. Getting the following error:
> 
> ```
> > Task :server:customModules:FDA-HPHI-UserRegWS:compileJava FAILED
> C:\dev\labkey\trunk\server\customModules\FDA-HPHI-UserRegWS\src\com\hphc\mystudies\FdahpUserRegWSManager.java:111: error: cannot find symbol
>             SchemaTableInfo table = (SchemaTableInfo)FdahpUserRegWsSchema.getInstance().getParticipantDetails();
>             ^
>   symbol:   class SchemaTableInfo
>   location: class FdahpUserRegWSManager
> ```
> I will keep everyone posted on this

This might be because 17.3 is an outdated version. I fixed this by, Instead of downloading labkey-client-api-17.3.jar, simply removing a line from the `build.gradle` file @formanojhr posted.
`compile name: 'labkey-client-api-17.3'`.

Resulting in the following file at `fdahpUserRegWS/build.gradle`:
```
import org.labkey.gradle.util.BuildUtils

apply plugin: 'java'
apply plugin: 'org.labkey.module'

repositories {
    flatDir {
        dirs 'lib'
    }
    mavenCentral()
}

sourceSets {
    engine {
        java {
            srcDirs = ['./enginesrc']
        }
    }
}

dependencies {
    compile name: 'apns-1.0.0.Beta6'
    compile group: 'org.labkey', name: 'labkey-client-api', version: '15.3.0'
    BuildUtils.addLabKeyDependency(project: project, config: "compile", depProjectPath: ":server:api")
    BuildUtils.addLabKeyDependency(project: project, config: "compile", depProjectPath: ":server:internal")
    engineCompile "org.labkey:labkey-client-api:${project.version}"
}
```

After doing this I was able to access the ping.api endpoint. However I am not sure if this can have unintended consequences later down the road, a better solution might be replacing "17.3" with a variable like ${labkeyVersion} or ${project.version}.

On an other note I also found it to be safe to do a search replace inside the fdahpUserRegWS folder.
Replacing `TableInfo` with `SchemaTableInfo`, this however is not equivalent to the solution @formanojhr posted since I am not doing any type casting.

**Comment by FDA-MyStudies** [comment]
Hi @zardarkhan ,
Thanks for your interest in the MyStudies system.  Apologies for the year-long delay in responding, but we only recently received additional support in order to be able to fully support the developer community with the MyStudies app.  In response to your original question:

The latest version of Labkey uses Gradle system to build project instead of Ant.  So checkout the latest version of project from the github in intellij IDE and start building.

- After importing the project, in your settings.gradle file, find the commented out line with this text:

 //include ":server:optionalModules:workflow"

Underneath this line, add these two lines:

include ":server:customModules:fdahpUserRegWS"
include ":server:customModules:fdahpUserRegWS:distributions:Registration"

- To deploy the app use the following command
gradlew cleanBuild deployApp

- Please refer the latest documentation for more information.

- WCP-WS and user-Reg should be build as a two individual projects
